rust/src/types/amount.rs

```diff
@@ -0,0 +1,582 @@
+        }
+        let idx = index[31];
+        self[idx as usize].into()
+    }
+}
+
+impl Shl for u256 {
+    type Output = Self;
+
+    fn shl(self, rhs: Self) -> Self::Output {
+        if rhs > u256::from(255u8) {
+            return u256::ZERO;
+        }
+        let value: U256 = self.into();
+        let shift = rhs[31] as u32;
+        (value.wrapping_shl(shift)).into()
+    }
+}
+
+impl Shr for u256 {
+    type Output = Self;
+
+    fn shr(self, rhs: Self) -> Self::Output {
+        if rhs > u256::from(255u8) {
+            return u256::ZERO;
+        }
+        let value: U256 = self.into();
+        let shift = rhs[31] as u32;
+        (value.wrapping_shr(shift)).into()
+    }
+}
+
+impl u256 {
+    pub fn sar(self, rhs: Self) -> Self {
+        let negative = self[0] & 0x80 > 0;
+        if rhs > u256::from(255u8) {
+            if negative {
+                return u256::MAX;
+            } else {
+                return u256::ZERO;
+            }
+        }
+        let value: U256 = self.into();
+        let shift = rhs[31] as u32;
+        let mut shr = value.wrapping_shr(shift);
+        if negative {
+            shr |= U256::MAX.wrapping_shl(255 - shift);
+        }
+        shr.into()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use bnum::{
+        cast::CastFrom,
+        types::{I256, U256, U512},
+    };
+
+    use crate::types::amount::u256;
+
+    #[test]
+    fn display() {
+        assert_eq!(
+            format!(
+                "{}",
+                u256::from([
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                    0, 0, 0, 0, 0, 0,
+                ])
+            ),
+            "0x0000000000000000_0000000000000000_0000000000000000_0000000000000000"
+        );
+        assert_eq!(
+            format!(
+                "{}",
+                u256::from([
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                    0, 0, 0, 0, 0, 0xfe,
+                ])
+            ),
+            "0x0000000000000000_0000000000000000_0000000000000000_00000000000000fe"
+        );
+        assert_eq!(
+            format!(
+                "{}",
+                u256::from([
+                    0xfe, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                    0, 0, 0, 0, 0, 0, 0,
+                ])
+            ),
+            "0xfe00000000000000_0000000000000000_0000000000000000_0000000000000000"
+        );
+    }
+
+    #[test]
+    fn conversions() {
+        let v1 = u256::from([
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 1,
+        ]);
+        let v255 = u256::from([
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 255,
+        ]);
+
+        assert_eq!(u256::from(false), u256::ZERO);
```

Comment:
Out of curiosity: is there also an `expect_eq` like in other test frameworks that would report an issue but continues with the test?

Reply:
I don't think so. But it would be possible to split it up into multiple tests and put the common setup code into a function which is not annotated with `#[test]`.